src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,354 @@
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pScheduler = std::make_unique<ControllerRenderingScheduler>(m_screenInfo);
+
+    connect(this,
+            &ControllerRenderingEngine::setupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::stopRequested,
+            this,
+            &ControllerRenderingEngine::finish);
+    connect(m_pScheduler.get(),
+            &ControllerRenderingScheduler::frameRequested,
+            this,
+            &ControllerRenderingEngine::renderFrame);
+
+    moveToThread(qApp->thread());
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.warning() << "The ControllerEngine is being deleted but hasn't been "
+                             "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pScheduler->isFinished() &&
+            !m_pScheduler->isInterruptionRequested()) {
+        kLogger.warning() << "Render thread has or ir about to terminate. Cannot "
+                             "start this render anymore.";
+        return;
+    }
+    m_pScheduler->start(QThread::NormalPriority);
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pScheduler && m_pScheduler->isRunning();
+}
+
+void ControllerRenderingEngine::requestSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
+    VERIFY_OR_DEBUG_ASSERT(QThread::currentThread() != thread()) {
+        kLogger.warning() << "Unable to setup OpenGL rendering context from the same "
+                             "thread as the render object";
+        return;
+    }
+    emit setupRequested(qmlEngine);
+
+    const auto lock = lockMutex(&m_mutex);
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+}
+
+void ControllerRenderingEngine::requestSend(Controller* controller, const QByteArray& frame) {
+    VERIFY_OR_DEBUG_ASSERT(m_pScheduler) {
+        return;
+    }
+    m_pScheduler->requestSend(controller, frame);
+}
+
+void ControllerRenderingEngine::setup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    QSurfaceFormat format;
+    format.setSamples(16);
+    format.setDepthBufferSize(16);
+    format.setStencilBufferSize(8);
+
+    const auto lock = lockMutex(&s_glMutex);
+
+    m_context = std::make_unique<QOpenGLContext>();
+    m_context->setFormat(format);
+    VERIFY_OR_DEBUG_ASSERT(m_context->create()) {
+        kLogger.warning() << "Unable to initialize controller screen rendering. Giving up";
+        m_waitCondition.wakeAll();
+        return;
+    }
+    connect(m_context.get(),
+            &QOpenGLContext::aboutToBeDestroyed,
+            this,
+            &ControllerRenderingEngine::finish,
+            Qt::BlockingQueuedConnection);
```

Comment:
Not in this commit, but they used to in the previous one. Will remove the connection flag in case we decide to keep the single threaded version

Reply:
Yes, we need to because if they're in the same thread, `BlockingQueuedConnection` would deadlock. However, if they in different threads, this needs very explicit documentation and synchronization...